Snowflake.Data/Core/Converter/JsonToStructuredTypeConverter.cs

```diff
@@ -0,0 +1,423 @@
+                && keyType != typeof(long) && keyType != typeof(long?))
+            {
+                throw new Exception($"Unsupported key type of dictionary {keyType}");
+            }
+            if (json == null || json.Type == JTokenType.Null || json.Type == JTokenType.Undefined)
+            {
+                return null;
+            }
+            if (json.Type != JTokenType.Object)
+            {
+                throw new Exception("Json is not an object");
+            }
+            if (fields == null || fields.Count != 2)
+            {
+                throw new Exception("Expected map to have 2 metadata fields");
+            }
+            var keyMetadata = fields[0];
+            var fieldMetadata = fields[1];
+            var jsonObject = (JObject)json;
+            var dictionaryType = type.IsAbstract ? typeof(Dictionary<,>).MakeGenericType(keyType, valueType) : type;
+            var result = (IDictionary) Activator.CreateInstance(dictionaryType);
+            using (var jsonEnumerator = jsonObject.GetEnumerator())
+            {
+                while (jsonEnumerator.MoveNext())
+                {
+                    var jsonPropertyWithValue = jsonEnumerator.Current;
+                    var fieldValue = jsonPropertyWithValue.Value;
+                    var key = IsTextMetadata(keyMetadata) || IsFixedMetadata(keyMetadata)
+                        ? ConvertToUnstructuredType(keyMetadata, keyType, jsonPropertyWithValue.Key)
+                        : throw new Exception("Could not recognize key type for map");
+                    var value = ConvertToStructuredOrUnstructuredValue(valueType, fieldMetadata, fieldValue);
+                    result.Add(key, value);
+                }
+            }
+            return result;
+        }
+
+        private static object ConvertToStructuredOrUnstructuredValue(Type valueType, FieldMetadata fieldMetadata, JToken fieldValue)
+        {
+            try
+            {
+                if (IsObjectMetadata(fieldMetadata) && IsStructured(fieldMetadata))
+                {
+                    return ConvertToObject(valueType, fieldMetadata.fields, fieldValue);
+                }
+
+                if (IsArrayMetadata(fieldMetadata) && IsStructured(fieldMetadata))
+                {
+                    var nestedType = GetNestedType(valueType);
+                    return ConvertToArray(valueType, nestedType, fieldMetadata.fields, fieldValue);
+                }
+
+                if (IsMapMetadata(fieldMetadata) && IsStructured(fieldMetadata))
+                {
+                    var keyValueTypes = GetMapKeyValueTypes(valueType);
+                    return ConvertToMap(valueType, keyValueTypes[0], keyValueTypes[1], fieldMetadata.fields, fieldValue);
+                }
+
+                return ConvertToUnstructuredType(fieldMetadata, valueType, fieldValue);
+            }
+            catch (Exception e)
+            {
+                if (e is SnowflakeDbException)
```

Comment:
next place with repackaging to Snowflake specific exception

Reply:
ok